plutus-core/plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -186,20 +186,67 @@ instance PrettyDefaultBy config Integer => PrettyBy config ExBudget where
         , "}"
         ]
 
-data ExBudgetState exBudgetCat = ExBudgetState
-    { _exBudgetStateTally  :: ExTally exBudgetCat -- ^ for counting what cost how much
-    , _exBudgetStateBudget :: ExBudget -- ^ for making sure we don't spend too much
-    }
+-- | @(-)@ on 'ExCPU'.
+minusExCPU :: ExCPU -> ExCPU -> ExCPU
+minusExCPU = coerce $ (-) @Integer
+
+-- | @(-)@ on 'ExMemory'.
+minusExMemory :: ExMemory -> ExMemory -> ExMemory
+minusExMemory = coerce $ (-) @Integer
+
+-- | Subtract an 'ExBudget' from an 'ExRestrictingBudget'.
+minusExBudget :: ExRestrictingBudget -> ExBudget -> ExRestrictingBudget
+ExRestrictingBudget (ExBudget cpuL memL) `minusExBudget` ExBudget cpuR memR =
+    ExRestrictingBudget $ ExBudget (cpuL `minusExCPU` cpuR) (memL `minusExMemory` memR)
+
+-- | A budgeting mode to execute an evaluator in.
+data ExBudgetMode
+    = Counting                         -- ^ For calculating the cost of execution.
+    | Tallying                         -- ^ For a detailed report on what costs how much +
+                                       --     the same overall budget that 'Counting' gives.
+    | Restricting ExRestrictingBudget  -- ^ For execution, to avoid overruns.
+
+-- | The state of budgeting. Gets initialized from an 'ExBudgetMode'.
+data ExBudgetState exBudgetCat
```

Comment:
I hesitate to suggest this, but should we do something like use a type family for the state type? The reason I suggest this is even if we took a `spendsBudget` function as an argument, we'd still have to branch on the type of state we have (presumably throwing if it wasn't the one we expected). So if we wanted to cut that branch from `spendsBudget` I think we'd need a type family for the state type?

Reply:
I think we don't, but anyway, that's for a different PR.